qpc/utils.py

```diff
@@ -520,3 +529,47 @@ def check_extension(extension, path):
     if extension not in path:
         print(t(messages.OUTPUT_FILE_TYPE % extension))
         sys.exit(1)
+
+
+def write_encryption_key_if_non_existent():
+    """Generate encryption key.
+
+    Key will be generated only once and saved at INSIGHTS_ENCRYPTION file,
+    the function will check its existence every time it is called
+    """
+    if not os.path.exists(INSIGHTS_ENCRYPTION):
+        key = Fernet.generate_key()
+        with open(INSIGHTS_ENCRYPTION, "wb") as key_file:
+            key_file.write(key)
+            os.chmod(INSIGHTS_ENCRYPTION, 0o600)
+
+
+def load_encryption_key():
+    """Load encryption from insights_encryption file."""
+    stats = os.stat(INSIGHTS_ENCRYPTION)
+    if oct(stats.st_mode).endswith("00"):
+        return open(INSIGHTS_ENCRYPTION, "rb").read()
+
+    raise QPCEncryptionKeyError(
+        "There was a problem while trying to load the password encryption key."
+    )
+
+
+def encrypt_password(password):
+    """Encrypt password before saving it to insights_login_config file."""
+    write_encryption_key_if_non_existent()
+    key = load_encryption_key()
+
+    encryption_algorithm = Fernet(key)
+
+    encrypted_password = encryption_algorithm.encrypt(password.encode())
+    return encrypted_password.decode()
+
+
+def decrypt_password(password):
```

Comment:
@bruno-fs Hey, you asked me to ass an exception here in case anything goes wrong with decryption, but I believe we should add to the function that calls read_insights_login_config instead. This will happen as part of the publish implementation. 
What do you and @gtanzillo think?

Reply:
I'll start by the end: yes, the place to properly treat the exception is the publish command.

but it shouldn't deal with cryptography exceptions, only QPCErrors. the only functios that know about python cryptography are the encryption/decryption related ones - and in my view is their responsibility to "convert" expected 3rd party exceptions to the ones qpc entrypoints expects

Reply:
I'm 👍 with what @bruno-fs said ^^. Agree that the exception handling for encrypt/decrypt should be at the place where those operations occur. They can always be handled and then bubbled up the call stack if it's necessary for the calles to also handle them

Reply:
I understand and agree with both of you. I'll push a fixup with those changes and after the publish pr and this one get merged I'll add a QPCError whenever read_insights_login_config is called.

Reply:
good call